File: src/main/java/kirjanpito/ui/javafx/dialogs/AccountSelectionDialogFX.java
package kirjanpito.ui.javafx.dialogs;

import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.collections.transformation.FilteredList;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.input.KeyCode;
import javafx.scene.layout.*;
import javafx.stage.Modality;
import javafx.stage.Stage;
import javafx.stage.Window;

import kirjanpito.db.Account;

import java.util.List;
import java.util.function.Consumer;

/**
 * JavaFX tilinvalintadialogi (F9).
 * 
 * Näyttää tilikartan ja mahdollistaa tilin valinnan hakusanalla.
 */
public class AccountSelectionDialogFX {
    
    private Stage dialog;
    private TextField searchField;
    private TableView<Account> accountTable;
    private ObservableList<Account> allAccounts;
    private FilteredList<Account> filteredAccounts;
    private Consumer<Account> onAccountSelected;
    private Account selectedAccount;
    private CheckBox showFavoritesOnly;
    
    public AccountSelectionDialogFX(Window owner) {
        dialog = new Stage();
        dialog.initModality(Modality.APPLICATION_MODAL);
        dialog.initOwner(owner);
        dialog.setTitle("Tilin valinta");
        dialog.setMinWidth(500);
        dialog.setMinHeight(400);
        
        createContent();
    }
    
    private void createContent() {
        VBox root = new VBox(10);
        root.setPadding(new Insets(16));
        root.setStyle("-fx-background-color: #ffffff;");
        
        // Search field
        HBox searchBox = new HBox(10);
        searchBox.setAlignment(Pos.CENTER_LEFT);
        
        Label searchLabel = new Label("Haku:");
        searchField = new TextField();
        searchField.setPromptText("Tilinumero tai nimi...");
        searchField.setPrefWidth(300);
        HBox.setHgrow(searchField, Priority.ALWAYS);
        
        showFavoritesOnly = new CheckBox("Vain suosikit");
        
        searchBox.getChildren().addAll(searchLabel, searchField, showFavoritesOnly);
        
        // Table
        accountTable = new TableView<>();
        accountTable.setColumnResizePolicy(TableView.CONSTRAINED_RESIZE_POLICY);
        VBox.setVgrow(accountTable, Priority.ALWAYS);
        
        TableColumn<Account, String> numberCol = new TableColumn<>("Numero");
        numberCol.setCellValueFactory(new PropertyValueFactory<>("number"));
        numberCol.setPrefWidth(80);
        numberCol.setMinWidth(60);
        
        TableColumn<Account, String> nameCol = new TableColumn<>("Nimi");
        nameCol.setCellValueFactory(new PropertyValueFactory<>("name"));
        nameCol.setPrefWidth(300);
        
        TableColumn<Account, String> typeCol = new TableColumn<>("Tyyppi");
        typeCol.setCellValueFactory(cellData -> {
            Account acc = cellData.getValue();
            String type = getAccountTypeName(acc.getType());
            return new javafx.beans.property.SimpleStringProperty(type);
        });
        typeCol.setPrefWidth(100);
        
        accountTable.getColumns().addAll(numberCol, nameCol, typeCol);
        
        // Buttons
        HBox buttonBox = new HBox(10);
        buttonBox.setAlignment(Pos.CENTER_RIGHT);
        buttonBox.setPadding(new Insets(10, 0, 0, 0));
        
        Button okButton = new Button("Valitse");
        okButton.setDefaultButton(true);
        okButton.setStyle("-fx-background-color: #2563eb; -fx-text-fill: white;");
        okButton.setOnAction(e -> selectAndClose());
        
        Button cancelButton = new Button("Peruuta");
        cancelButton.setCancelButton(true);
        cancelButton.setOnAction(e -> dialog.close());
        
        buttonBox.getChildren().addAll(cancelButton, okButton);
        
        root.getChildren().addAll(searchBox, accountTable, buttonBox);
        
        // Event handlers
        setupEventHandlers();
        
        Scene scene = new Scene(root, 550, 450);
        dialog.setScene(scene);
    }
    
    private void setupEventHandlers() {
        // Search filter
        searchField.textProperty().addListener((obs, oldVal, newVal) -> {
            filterAccounts(newVal);
        });
        
        // Favorites filter
        showFavoritesOnly.selectedProperty().addListener((obs, oldVal, newVal) -> {
            filterAccounts(searchField.getText());
        });
        
        // Double-click to select
        accountTable.setOnMouseClicked(event -> {
            if (event.getClickCount() == 2) {
                selectAndClose();
            }
        });
        
        // Enter to select
        accountTable.setOnKeyPressed(event -> {
            if (event.getCode() == KeyCode.ENTER) {
                selectAndClose();
            } else if (event.getCode() == KeyCode.ESCAPE) {
                dialog.close();
            }
        });
        
        // Search field Enter moves to table
        searchField.setOnKeyPressed(event -> {
            if (event.getCode() == KeyCode.ENTER || event.getCode() == KeyCode.DOWN) {
                if (!filteredAccounts.isEmpty()) {
                    accountTable.getSelectionModel().selectFirst();
                    accountTable.requestFocus();
                }
            } else if (event.getCode() == KeyCode.ESCAPE) {
                dialog.close();
            }
        });
    }
    
    private void filterAccounts(String search) {
        if (filteredAccounts == null) return;
        
        final String searchLower = search != null ? search.toLowerCase().trim() : "";
        final boolean favoritesOnly = showFavoritesOnly.isSelected();
        
        filteredAccounts.setPredicate(account -> {
            // Favorites filter (flag bit 0 = favorite)
            if (favoritesOnly && (account.getFlags() & 1) == 0) {
                return false;
            }
            
            // Search filter
            if (searchLower.isEmpty()) {
                return true;
            }
            
            // Match by number prefix
            if (account.getNumber().startsWith(searchLower)) {
                return true;
            }
            
            // Match by name
            if (account.getName().toLowerCase().contains(searchLower)) {
                return true;
            }
            
            return false;
        });
        
        // Select first match
        if (!filteredAccounts.isEmpty()) {
            accountTable.getSelectionModel().selectFirst();
        }
    }
    
    private void selectAndClose() {
        selectedAccount = accountTable.getSelectionModel().getSelectedItem();
        if (selectedAccount != null && onAccountSelected != null) {
            onAccountSelected.accept(selectedAccount);
        }
        dialog.close();
    }
    
    private String getAccountTypeName(int type) {
        switch (type) {
            case 1: return "Vastaavaa";
            case 2: return "Vastattavaa";
            case 3: return "Tulos";
            case 4: return "Oma pääoma";
            case 5: return "Edellisten voitto";
            default: return "";
        }
    }
    
    public void setAccounts(List<Account> accounts) {
        allAccounts = FXCollections.observableArrayList(accounts);
        filteredAccounts = new FilteredList<>(allAccounts, p -> true);
        accountTable.setItems(filteredAccounts);
    }
    
    public void setOnAccountSelected(Consumer<Account> callback) {
        this.onAccountSelected = callback;
    }
    
    public void setInitialSearch(String search) {
        if (search != null && !search.isEmpty()) {
            searchField.setText(search);
            filterAccounts(search);
        }
    }
    
    public Account getSelectedAccount() {
        return selectedAccount;
    }
    
    public void show() {
        // Focus search field
        Platform.runLater(() -> {
            searchField.requestFocus();
            searchField.selectAll();
        });
        
        dialog.showAndWait();
    }
    
    /**
     * Näyttää dialogin ja palauttaa valitun tilin.
     */
    public static Account showAndSelect(Window owner, List<Account> accounts, String initialSearch) {
        AccountSelectionDialogFX dialog = new AccountSelectionDialogFX(owner);
        dialog.setAccounts(accounts);
        if (initialSearch != null) {
            dialog.setInitialSearch(initialSearch);
        }
        dialog.show();
        return dialog.getSelectedAccount();
    }
}

File: src/main/kotlin/kirjanpito/ui/javafx/dialogs/AccountSelectionDialogFX.kt
package kirjanpito.ui.javafx.dialogs

import javafx.application.Platform
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.collections.transformation.FilteredList
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Parent
import javafx.scene.control.*
import javafx.scene.control.cell.PropertyValueFactory
import javafx.scene.input.KeyCode
import javafx.scene.layout.HBox
import javafx.scene.layout.Priority
import javafx.scene.layout.VBox
import javafx.stage.Window
import kirjanpito.db.Account

/**
 * JavaFX tilinvalintadialogi (F9).
 * 
 * Näyttää tilikartan ja mahdollistaa tilin valinnan hakusanalla.
 * 
 * Migrated from Java to Kotlin for better code conciseness and null-safety.
 */
class AccountSelectionDialogFX(owner: Window?) : BaseDialogFX(
    owner = owner,
    title = "Tilin valinta",
    width = 550.0,
    height = 450.0,
    resizable = true
) {
    
    private var searchField: TextField? = null
    private var accountTable: TableView<Account>? = null
    private var allAccounts: ObservableList<Account>? = null
    private var filteredAccounts: FilteredList<Account>? = null
    private var onAccountSelected: ((Account) -> Unit)? = null
    private var selectedAccount: Account? = null
    private var showFavoritesOnly: CheckBox? = null
    
    override fun hasButtons(): Boolean = true
    
    override fun createOKButton(): Button {
        return Button("Valitse").apply {
            isDefaultButton = true
            style = "-fx-background-color: #2563eb; -fx-text-fill: white; -fx-min-width: 80;"
            setOnAction {
                if (onOK()) {
                    dialog.close()
                }
            }
        }
    }
    
    override fun createContent(): Parent {
        val root = VBox(10.0).apply {
            padding = Insets(16.0)
            style = "-fx-background-color: #ffffff;"
        }
        
        // Search field
        val searchBox = HBox(10.0).apply {
            alignment = Pos.CENTER_LEFT
        }
        
        val searchLabel = Label("Haku:")
        searchField = TextField().apply {
            promptText = "Tilinumero tai nimi..."
            prefWidth = 300.0
            HBox.setHgrow(this, Priority.ALWAYS)
        }
        
        showFavoritesOnly = CheckBox("Vain suosikit")
        
        searchBox.children.addAll(searchLabel, searchField, showFavoritesOnly)
        
        // Table
        accountTable = TableView<Account>().apply {
            columnResizePolicy = TableView.CONSTRAINED_RESIZE_POLICY
            VBox.setVgrow(this, Priority.ALWAYS)
        }
        
        val numberCol = TableColumn<Account, String>("Numero").apply {
            cellValueFactory = PropertyValueFactory("number")
            prefWidth = 80.0
            minWidth = 60.0
        }
        
        val nameCol = TableColumn<Account, String>("Nimi").apply {
            cellValueFactory = PropertyValueFactory("name")
            prefWidth = 300.0
        }
        
        val typeCol = TableColumn<Account, String>("Tyyppi").apply {
            cellValueFactory = javafx.util.Callback<TableColumn.CellDataFeatures<Account, String>, javafx.beans.value.ObservableValue<String>> { cellData ->
                val account = cellData.value
                javafx.beans.property.SimpleStringProperty(getAccountTypeName(account.type))
            }
            prefWidth = 100.0
        }
        
        accountTable?.columns?.addAll(numberCol, nameCol, typeCol)
        
        root.children.addAll(searchBox, accountTable)
        
        // Event handlers
        setupEventHandlers()
        
        return root
    }
    
    private fun setupEventHandlers() {
        val search = searchField ?: return
        val table = accountTable ?: return
        val favorites = showFavoritesOnly ?: return
        
        // Search filter
        search.textProperty().addListener { _, _, newVal ->
            filterAccounts(newVal)
        }
        
        // Favorites filter
        favorites.selectedProperty().addListener { _, _, _ ->
            filterAccounts(search.text)
        }
        
        // Double-click to select
        table.setOnMouseClicked { event ->
            if (event.clickCount == 2) {
                selectAndClose()
            }
        }
        
        // Enter to select, Escape to close
        table.setOnKeyPressed { event ->
            when (event.code) {
                KeyCode.ENTER -> selectAndClose()
                KeyCode.ESCAPE -> dialog.close()
                else -> {}
            }
        }
        
        // Search field Enter/Down moves to table
        search.setOnKeyPressed { event ->
            when (event.code) {
                KeyCode.ENTER, KeyCode.DOWN -> {
                    if (!filteredAccounts.isNullOrEmpty()) {
                        table.selectionModel.selectFirst()
                        table.requestFocus()
                    }
                }
                KeyCode.ESCAPE -> dialog.close()
                else -> {}
            }
        }
    }
    
    private fun filterAccounts(search: String?) {
        val filtered = filteredAccounts ?: return
        val favorites = showFavoritesOnly ?: return
        
        val searchLower = search?.lowercase()?.trim() ?: ""
        val favoritesOnly = favorites.isSelected
        
        filtered.setPredicate { account: Account ->
            // Favorites filter (flag bit 0 = favorite)
            if (favoritesOnly && (account.flags and 1) == 0) {
                return@setPredicate false
            }
            
            // Search filter
            if (searchLower.isEmpty()) {
                return@setPredicate true
            }
            
            // Match by number prefix
            if (account.number.startsWith(searchLower, ignoreCase = true)) {
                return@setPredicate true
            }
            
            // Match by name
            account.name.lowercase().contains(searchLower)
        }
        
        // Select first match
        if (filtered.isNotEmpty()) {
            accountTable?.selectionModel?.selectFirst()
        }
    }
    
    private fun selectAndClose() {
        val table = accountTable ?: return
        selectedAccount = table.selectionModel.selectedItem
        selectedAccount?.let { account ->
            onAccountSelected?.invoke(account)
        }
        dialog.close()
    }
    
    private fun getAccountTypeName(type: Int): String {
        return when (type) {
            1 -> "Vastaavaa"
            2 -> "Vastattavaa"
            3 -> "Tulos"
            4 -> "Oma pääoma"
            5 -> "Edellisten voitto"
            else -> ""
        }
    }
    
    fun setAccounts(accounts: List<Account>) {
        allAccounts = FXCollections.observableArrayList(accounts)
        filteredAccounts = FilteredList(allAccounts!!) { true }
        accountTable?.items = filteredAccounts
    }
    
    fun setOnAccountSelected(callback: (Account) -> Unit) {
        this.onAccountSelected = callback
    }
    
    fun setInitialSearch(search: String?) {
        if (!search.isNullOrEmpty()) {
            searchField?.text = search
            filterAccounts(search)
        }
    }
    
    fun getSelectedAccount(): Account? = selectedAccount
    
    override fun onOK(): Boolean {
        selectAndClose()
        return true
    }
    
    override fun show() {
        // Focus search field
        Platform.runLater {
            searchField?.requestFocus()
            searchField?.selectAll()
        }
        
        super.show()
    }
    
    companion object {
        /**
         * Näyttää dialogin ja palauttaa valitun tilin.
         */
        @JvmStatic
        fun showAndSelect(owner: Window?, accounts: List<Account>, initialSearch: String?): Account? {
            val dialog = AccountSelectionDialogFX(owner)
            dialog.setAccounts(accounts)
            if (initialSearch != null) {
                dialog.setInitialSearch(initialSearch)
            }
            dialog.showAndWait()
            return dialog.getSelectedAccount()
        }
    }
}
